### ISSUE TITLE
Prevent users from seeing personal data of persons outside their jursisdiction (pseudonymization) [8]

### ISSUE THREAD

**Issue by MateStrysewskeSym** [hzi-braunschweig/SORMAS-Project#1158]
### Situation Description
Currently users can see all data fields of each entity they can access (e.g. cases and contacts). In some countries this is not allowed for personal data like first and last names, the full birth date (age is ok) and the address.

### Feature Description
Introduce a new user right to SORMAS that defines whether a user is allowed to see personal data of persons outside the user's jurisdication. This would for example be relevant for a case the user can see because of having access to a contact of that case.
In addition a user right is needed to define this for persons inside the jurisdiction of a user. E.g. a national user may not be allowed to see any personal data.

Follow-up issue: #2130

**Comment by MateStrysewskeSym** [comment]
NEW USER RIGHT: SEE_PERSONAL_DATA_IN_JURISDICTION & SEE_PERSONAL_DATA_OUTSIDE_JURISDICTION
=> Prevent users WITHOUT this right from seeing:
  * First name and Last name
  * Day of birth
  * All person address fields except region and district
  * Case community, health facility and point of entry
  * Reporting user

NEW USER RIGHT: SEE_SENSITIVE_DATA_OUTSIDE_JURISDICTION
=> Prevent users WITHOUT this right from seeing:
  * Tasks assigned to users on district level or below
  * User names of personnel on district level or below
  * Any location information below district level (e.g. the facility of a case)
  * Names of laboratories where samples have been sent to
  * All fields that allow users to type in free text
  * Detailed information about the employee status of cases or contacts

How to do this technically?

* Add new user rights
* Add new annotations to fields: @PersonalData & @SensitiveData
* Use this new annotation when using isVisibleAllowed and associated methods
* Adjust userFilter for tasks to hide those that users aren't allowed to see
* Hide columns/data from directories/lists

**Comment by markusmann-vg** [comment]
The pseudonymization has to take place in the backend as we must not transmit sensitive data through REST.

**Comment by leventegal-she** [comment]
Implementation details for personal fields
- [ ] set personal data to null DTO-s after reading from the DB
- [ ] make sure backend doesn't allow update data with pseudonymized fields
- [ ] disable edit on UI(web and android) for the whole object that has pseudonymized fields: using a boolean flag on the DTO(s)

Open points to clarify further
- [ ] peseudonymization of reporting user
- [ ] peseudonymization of personal data in events

**Comment by MateStrysewskeSym** [comment]
New information:

* Reporting user needs to be pseudonymised for every entity, same as personal data in events
* Pseudonymised fields need to be made read-only independent of the read-only state of the rest of the form; users on region or national level might still be able to edit cases, contacts etc. that contain pseudonymised data

**Comment by leventegal-she** [comment]
Implemented pseudonymization of personal data

**Comment by MateStrysewskeSym** [comment]
Things that still need discussion:

* How do we handle search fields? In theory, users can still search for cases by person name which leaks personal data
* What about the merge duplicate cases view? It still displays the first and last name and would be pretty unusable if it didn't

**Comment by MateStrysewskeSym** [comment]
- [ ] Refactor VisitFacadeEjb.getVisitsExportList; it's not an option to execute an additional query per VisitExportDto to get the contacts because this has a massively negative impact on performance. This needs to be done before the release.

**Comment by MateStrysewskeSym** [comment]
The code currently does three query per person to check whether it's in the jurisdiction of the user. This needs to be refactored.
